### PR TITLE
Non-modal dialogs should remain in the parent browseMode treeInterceptor

### PR DESCRIPTION
**Pull request by michaelDCurran**
### Link to issue number:
None.

### Summary of the issue:
The aria-modal html attribute was added to ARIA 1.1 to state whether a dialog / alertdialog is modal or not. Modal means that the user cannot interact with anything else in the application while the dialog is currently visible. 
Currently NVDA treets all web dialogs as modal as it  gives the dialog a new browseMode document, limited to that dialog, and also does not render any of the dialog's content in the existing parent browseMode document.
To conform with aria-modal, it should only do this if aria-modal is true, and otherwise keep the dialog as part of the parent browseMode document.

### Description of how this pull request fixes the issue:
* MSHTML NVDAObject: map aria-modal="true"  to NVDA's STATE_MODAL.
* MSHTML vbufBackend: only refuse to render the content of a dialog if the dialog node has aria-modal="true".
* Gecko/Chrome vbufBackend: only refuse to render the content of a dialog if the dialog node has aria-modal="true".
* No longer set focus to dialogs in browseMode when pressing enter on them if they do not have the modal state. 
* WebDialog NVDAObject behaviour: shouldCreateTreeInterceptor now refuses to create a treeInterceptor for a dialog if it does not have the modal state.

In short, now if a web dialog is not modal, all of its content will appear in the existing browseMode document. However, if the dialog is modal, then  its content will not be accessible in the existing browseMode document, and a new browseMode document limited to the dialog will be used. This was the previous behaviour for all web dialogs.

### Testing performed:
Using the html testcase: 
[dialog-modal-modeless.html.txt](https://github.com/nvaccess/nvda/files/1930487/dialog-modal-modeless.html.txt)

Tested modal and non-modal dialogs in Chrome, Firefox and Internet Explorer.
 1. Open the testcase in the browser
2. in BrowseMode arrow to the modal dialog button and activate it.
3. Ensure that NVDA reads the content of the dialog.
4. Ensure that it is impossible to arrow outside of this dialog in browseMode.
5. Dismiss the dialog by pressing Ok.
6. Activate the modeless dialog button.
7. Ensure NVDA reads the content of the dialog.
8. Ensure that arrowing up and down can move out of the dialog content and down to the original dialog buttons.

### Known issues with pull request:
Behaviour for any web dialog that does not specify aria-modal at all will now act as though aria-modal="false". It previously would have acted like aria-modal="true". However, this seems to be in line with the ARIA specification. Web authors should explicitly apply aria-modal to be clar what they want in future.
  
### Change log entry:
Changes:
Some dialogs on the web that are not specifically treeted as modal dialogs now appear in the same browseMode document as the surrounding web document, rather than getting their own browseMode document.

### CODE REVIEW

**Comment by leonardder** [comment]
This pr bothers me a little. I do understand why this change might be necessary and according to the Aria specification, but I'm afraid that this will cause many modal dialogs not to be treated as such, just because the authors don't follow Aria 1.1.

**Comment by leonardder** [comment]
What will the default behavior for alertdialog be with this pr?

An alternative approach could be allowing quick navigation to embedded dialogs as well as treating them as containers, so you can quickly pass through them.

**Comment by ironcross32** [comment]
I have some concerns about this. At first glance, it seems OK, but delving under the surface, I can spot problems especially if the website is previously unfamiliar. Let's say a user opens an account with a new bank, they then log onto that bank's website and enroll for the online account. The account is created, now the user can log in and see all their standings with that bank. If the bank's site improperly, or indeed not at all, specified the ARIA attribute, a scenario like the one I will describe may occur. Let's say that the user wishes to transfer $300 from a savings account to a checking account of the same bank. They open the dialog, it is modal, but the ARIA attribute is not specified, so the user does not know it is modal. They fill in the relevant fields then scroll through looking for the button to submit and process the transfer, they find a button that reads, 'make transfer'. They activate it, and now all the information is gone. They navigate to the form again and the information isn't there. Maybe they assume the transfer took place, but it didn't. Now they have several bills which will be taken from this account, and insufficient funds to pay them.

Does this press the edges, yes, but the point is that there are users of all levels of skill who use NVDA, and while I would hazard that the overwhelming majority of us would investigate, would know we did not find the submit button, but lost focus on the dialog and ended up hitting the button to access that dialog once again, you just never can tell. I don't want this to serve as a recommendation to start watering down NVDA's feature set, I do think it should have features for all levels of users, however, as previously stated, I do have some concerns about this.

**Comment by BDH6456** [comment]
Either way I think there should be a sound for modal dialogs, as there is for switching between browse and focus modes. A similar change in navigation method is expected in a modal dialog, so a sound seems logical. I speak as one who has found myself dealing with a similar situation to that mentioned by @ironcross32.

**Comment by ultrasound1372** [comment]
I can agree here, what I would think is treat it via speech as a container, like it does frames. When entering a non-modal dialog you would speak something like dialog, and when you exit it it would go out of dialog, and container nav gestures would apply. I think that would be the most fluid way to have this integrate as normal. I don't necessarily know about a sound, sounds are another debate entirely, because if I start talking about sounds I'll get into those new sounds that got killed and yeah... Maybe have an option in document formatting to control the default behavior, default is to merge, you can check it to modalize it. It would be like "always treat dialogs as full forground" or something.

**Comment by jscholes** [comment]
> An alternative approach could be allowing quick navigation to embedded dialogs as well as treating them as containers, so you can quickly pass through them.

If dialog content is added to existing browse mode documents, providing ways to identify those sections of the page, jump directly to them and skip past them is imperative.

**Comment by derekriemer** [comment]
[**@jscholes**](https://github.com/jscholes) commented on [Apr 21, 2018, 7:47 PM MDT](https://github.com/nvaccess/nvda/pull/8194#issuecomment-383348081 "2018-04-22T01:47:17Z - Replied by Github Reply Comments"):
> > An alternative approach could be allowing quick navigation to embedded dialogs as well as treating them as containers, so you can quickly pass through them.
> 
> If dialog content is added to existing browse mode documents, providing ways to identify those sections of the page, jump directly to them and skip past them is imperative.

yes

**Comment by ultrasound1372** [comment]
I agree with @jscholes that definitely should be a thing. Maybe reat them as frames as far as keystrokes, we've used up almost the entirety of the keyboard for quick letter nav.

**Comment by michaelDCurran** [comment]
So firstly, yes, we can make dialogs speak like frames in that  it will 
say "out of dialog" at the bottom.
Secondly, I think you can already press o / shift+o to jump to dialogs.

**Comment by michaelDCurran** [comment]
@ironcross32: I am unclear as to whether you feel it is better for 
dialogs to be in their own browseMode document, or to be in the parent one.

**Comment by ironcross32** [comment]
@MichaelDCurran Most likely a brows mode document, but I am not a developer, and don't fully know the ramifications of it being in the parent window. Honestly, it could be either way, if as others have said, there is clear indication when you pass out of and into a dialog, that's fine. What I would not want to see happen is a user pressing buttons on a page that is outside the dialog they're trying to interact with. Visually, most modal dialogs will dim the content underneath, and clicking outside it will have no effect, but screen readers can navigate outside if the ARIA implementation is not there, or is poorly done.

So, to sum up, to me it doesn't matter, as long as I know where I am in the document, whether that be because I cannot navigate further because its a separate instance, or because I'm told that I've navigated past the dialog.

File: source/browseMode.py [hunk 1]
```diff
@@ -247,7 +247,17 @@ def _getLabelForProperties(self, labelPropertyGetter):
 class BrowseModeTreeInterceptor(treeInterceptorHandler.TreeInterceptor):
 	scriptCategory = inputCore.SCRCAT_BROWSEMODE
 	disableAutoPassThrough = False
-	APPLICATION_ROLES = (controlTypes.ROLE_APPLICATION, controlTypes.ROLE_DIALOG)
+
+	def _isApplicationObject(self,obj):
+		"""
+		Checks whether the given object in this browseMode treeInterceptor is an application (I.e. although the root of the object is represented in the treeInterceptor, its content is not).
```
Comment (feerrenrut) [inline source/browseMode.py]: Not super important, but this line could be split up.

**Review by feerrenrut — APPROVED** [review]
Looks good. Though, if the decision is made to also report leaving a modal, I think that change should be included with this.

**Comment by jcsteh** [comment]
I'd argue the default for aria-modal should be true for backwards compatability, but I do understand that's a spec argument, not an implementation argument.

I do worry about efficiency here. We're struggling with web apps becoming ever more cluttered, and this adds even more content to that. Consider popping up a list of keyboard shortcuts or the Gmail message composer, both of which are non-modal dialogs. There is potentially a huge amount of content in these dialogs that will now get blurted into the parent document. Yes, we could make them skippable containers, but that's still not as natural as, say, ctrl+end or ctrl+home for something which is clearly a somewhat separate entity.

It's also worth noting that in the existing implementation, even though we restrict navigation within dialogs, you can still get to the parent document with NVDA+ctrl+space. This is somewhat akin to being able to alt+tab out of a Windows modeless dialog or f6 out of a pop-up alert. Of course, this is not widely known, but I wonder whether we need to consider fixing that somehow as an alternative.

**Comment by leonardder** [comment]
An additional way to diverge between modal and non-modal dialogs, could be adding nvda+alt+tab to switch between tree interceptors. For example:

1. There is a parent document with a non-modal dialog in it
2. Both the non-modal dialog and the parent document have a tree interceptor
3. where control+nvda+space will bring you back from the dialog to the parent document, nvda+alt+tab could switch between the non-modal and the parent document, just like what you might expect from alt+tab or control+tab in a desktop application
